File: DynamicPrograming/java/MaxPalinSubsequence.java
import java.util.HashMap;
import java.util.Map;

public final class MaxPalinSubsequence {
  private MaxPalinSubsequence() {
  }

  public static int longest(String s) {
    if (s.isEmpty()) {
      return 0;
    }
    Map<String, Integer> memo = new HashMap<>();
    return dfs(0, s.length() - 1, s, memo);
  }

  private static int dfs(int i, int j, String s, Map<String, Integer> memo) {
    if (i == j) {
      return 1;
    }
    if (i > j) {
      return 0;
    }
    String key = i + "," + j;
    if (memo.containsKey(key)) {
      return memo.get(key);
    }
    int result;
    if (s.charAt(i) == s.charAt(j)) {
      result = 2 + dfs(i + 1, j - 1, s, memo);
    } else {
      int left = dfs(i + 1, j, s, memo);
      int right = dfs(i, j - 1, s, memo);
      result = Math.max(left, right);
    }
    memo.put(key, result);
    return result;
  }
}

File: DynamicPrograming/ts/maxPalinSubsequence.ts
export function maxPalinSubsequence(s: string): number {
  const memo = new Map<string, number>();

  const dfs = (i: number, j: number): number => {
    if (i === j) return 1;
    if (i > j) return 0;
    const key = `${i},${j}`;
    if (memo.has(key)) return memo.get(key)!;

    let result: number;
    if (s[i] === s[j]) {
      result = 2 + dfs(i + 1, j - 1);
    } else {
      result = Math.max(dfs(i + 1, j), dfs(i, j - 1));
    }
    memo.set(key, result);
    return result;
  };

  if (!s.length) return 0;
  return dfs(0, s.length - 1);
}
